toffy/normalize.py

```diff
@@ -184,69 +246,311 @@ def combine_tuning_curve_metrics(dir_list):
+                         norm_func_path=os.path.join('..', 'toffy', 'norm_func.json')):
+    """Normalizes image data based on median pulse height from the run and a tuning curve
+
+    Args:
+        img_dir (str): directory with the image data
+        norm_dir (str): directory where the normalized images will be saved
+        pulse_height_dir (str): directory containing per-fov pulse heights
+        panel_info (pd.DataFrame): mapping between channels and masses
+        mass_obj_func (str): class of function to use for modeling MPH over time per mass
+        extreme_vals (tuple): determines the range for norm vals which will raise a warning
+        norm_func_path (str): file containing the saved weights for the normalization function
+    """
+
+    # error checks
+    if not os.path.exists(norm_func_path):
+        raise ValueError("No normalization function found. You will need to run "
+                         "section 3 of the 1_set_up_toffy.ipynb notebook to generate the "
+                         "necessary function before you can normalize your data")
+
+    # create normalization function for mapping MPH to counts
+    with open(norm_func_path, 'r') as cf:
+        norm_json = json.load(cf)
+
+    img_fovs = io_utils.list_folders(img_dir, 'fov')
+
+    norm_weights, norm_name = norm_json['weights'], norm_json['name']
+    norm_func = create_prediction_function(norm_name, norm_weights)
 
-        for idx, chan in enumerate(channels):
-            io.imsave(os.path.join(output_fov_dir, chan + '.tiff'),
-                      normalized_images[0, :, :, idx], check_contrast=False)
+    # combine pulse heights together into single df
+    pulse_height_df = create_fitted_pulse_heights_file(pulse_height_dir=pulse_height_dir,
+                                                       panel_info=panel_info, norm_dir=norm_dir,
+                                                       mass_obj_func=mass_obj_func)
+    # add channel name to pulse_height_df
+    renamed_panel = panel_info.rename({'Mass': 'mass'}, axis=1)
+    pulse_height_df = pulse_height_df.merge(renamed_panel, how='left', on=['mass'])
+    pulse_height_df = pulse_height_df.sort_values('Target')
+
+    # make sure FOVs used to construct tuning curve are same ones being normalized
+    pulse_fovs = np.unique(pulse_height_df['fov'])
+    misc_utils.verify_same_elements(image_data_fovs=img_fovs, pulse_height_csv_files=pulse_fovs)
+
+    # loop over each fov
+    for fov in img_fovs:
+        # compute per-mass normalization constant
+        pulse_height_fov = pulse_height_df.loc[pulse_height_df['fov'] == fov, :]
+        norm_vals = norm_func(pulse_height_fov['pulse_height_fit'].values)
+        channels = pulse_height_fov['Target'].values
+
+        # get images
+        images = load_utils.load_imgs_from_tree(img_dir, fovs=[fov], channels=channels,
+                                                dtype='float32', img_sub_folder=img_sub_folder)
```

Comment:
Depending on how fast/slow this is, we might want to batch load the fovs, and then loop over each batch.